sky/utils/kubernetes_utils.py

```diff
@@ -578,3 +621,322 @@ def from_resources(cls,
+                               lambda c: c.type == 'ContainersReady')
+        if cont_ready_cond and \
+            cont_ready_cond.status == 'False':
+            # The main container is not ready. To be on the safe side
+            # and prevent a dangling sshjump pod, lets remove it and
+            # the service. Otherwise main container is ready and its lifecycle
+            # management script takes care of the cleaning.
+            kubernetes.core_api().delete_namespaced_pod(sshjump_name, namespace)
+            kubernetes.core_api().delete_namespaced_service(
+                sshjump_name, namespace)
+    # only warn and proceed as usual
+    except kubernetes.api_exception() as e:
+        logger.warning(f'Tried to check sshjump pod {sshjump_name},'
+                       f' but got error {e}\n. Consider running `kubectl '
+                       f'delete pod {sshjump_name} -n {namespace}` to manually '
+                       'remove the pod if it has crashed.')
+        # We encountered an issue while checking sshjump pod. To be on
+        # the safe side, lets remove its service so the port is freed
+        try:
+            kubernetes.core_api().delete_namespaced_service(
+                sshjump_name, namespace)
+        except kubernetes.api_exception():
+            pass
+
+
+def fill_sshjump_template(ssh_key_secret: str, sshjump_image: str,
+                          sshjump_name: str, service_type: str) -> Dict:
+    template_path = os.path.join(sky.__root_dir__, 'templates',
+                                 'kubernetes-sshjump.yml.j2')
+    if not os.path.exists(template_path):
+        raise FileNotFoundError(
+            'Template "kubernetes-sshjump.j2" does not exist.')
+    with open(template_path) as fin:
+        template = fin.read()
+    j2_template = jinja2.Template(template)
+    cont = j2_template.render(name=sshjump_name,
+                              image=sshjump_image,
+                              secret=ssh_key_secret,
+                              service_type=service_type)
+    content = yaml.safe_load(cont)
+    return content
+
+
+def check_port_forward_mode_dependencies() -> None:
+    """Checks if 'socat' and 'lsof' is installed"""
+    for name, option in [('socat', '-V'), ('lsof', '-v')]:
+        try:
+            subprocess.run([name, option],
+                           stdout=subprocess.DEVNULL,
+                           stderr=subprocess.DEVNULL,
+                           check=True)
+        except FileNotFoundError:
+            with ux_utils.print_exception_no_traceback():
+                raise RuntimeError(
+                    f'`{name}` is required to setup Kubernetes cloud with '
+                    f'`{KubernetesNetworkingMode.PORTFORWARD.value}` default '
+                    'networking mode and it is not installed. '
+                    'For Debian/Ubuntu system, install it with:\n'
+                    f'  $ sudo apt install {name}') from None
```

Comment:
Can we add the method for installing `socat` on mac, as `socat` seems required for local machine?
```
brew install socat
```